### PR TITLE
Extract Parameters and improve Introspection construction

### PR DESCRIPTION
**Pull request by tjhei**
- put Parameters in a separate .h/.cc
- Introspection is initialized from Parameters
- simplify Introspection construction

### CODE REVIEW

**Comment by tjhei** [comment]
Any comments @bangerth ? Thanks.

**Comment by bangerth** [comment]
I'm at J-L's workshop till tomorrow. Give me time till next week to look at stuff.

**Comment by tjhei** [comment]
I cleaned the code up some more. Any comments, @bangerth ?

**Comment by tjhei** [comment]
@bangerth ?

File: include/aspect/simulator.h [hunk 1]
```diff
@@ -95,248 +95,12 @@ namespace aspect
...
-          linear_momentum_x = 0x8,
-          linear_momentum_y = 0x10,
-          linear_momentum_z = 0x20,
-          linear_momentum   = 0x8+0x10+0x20,
-          net_rotation      = 0x40,
-          angular_momentum  = 0x80
-        };
-      };
+      typedef typename Parameters<dim>::NonlinearSolver NonlinearSolver;
+      typedef typename Parameters<dim>::NullspaceRemoval NullspaceRemoval;
```
Comment (bangerth) [inline include/aspect/simulator.h]: Do you do this for convenience, of for backward compatibility? In the latter case, maybe deprecate it?

**Comment by bangerth** [comment]
Other than the one question about the typedefs, the only thing I'm not in love with is the fact that you allocate the base elements and keep them till forever. Could we do something like the trick discussed in the documentation of the FESystem constructor you are calling? This way we don't have to keep two copied of everything around...

File: include/aspect/simulator.h (hunk 1 above)
Comment (tjhei) [inline include/aspect/simulator.h]: we basically had the same before:
```
 typedef typename NonlinearSolver::Kind NonlinearSolverKind;
```
but I guess we can get rid of this. Will do...

**Comment by bangerth** [comment]
OK, merge at your discretion then.

**Comment by tjhei** [comment]
I looked at the typedefs and this is not for backwards compatibility but to make the code more readable that uses the enums:
```
if (parameters.nullspace_removal & NullspaceRemoval::net_rotation)
```
is much nicer than
```
if (parameters.nullspace_removal & Parameters<dim>::NullspaceRemoval::net_rotation)
```
because it is used in many spots.

I opted for a simpler solution to delete the FEs (the data is around 45kb in 3d). Okay?

I cleaned up some other stuff.

File: include/aspect/introspection.h [hunk 2]
```diff
@@ -84,12 +107,29 @@ namespace aspect
...
+      const BlockIndices block_indices;
+
+      /**
        * A structure that contains FEValues extractors for every block of the
        * finite element used in the overall description.
        */
       struct Extractors
       {
-        Extractors (const unsigned int n_compositional_fields);
+        Extractors (const Introspection<dim>::ComponentIndices &component_indices);
```
Comment (bangerth) [inline include/aspect/introspection.h]: I think that if you write it like this, C++98 requires a `typename`. But, I also think that the members of the parent are accessible without qualification, so you could just write `ComponentIndices` without the prefix qualification.

File: source/simulator/introspection.cc [hunk 3]
```diff
@@ -22,151 +22,188 @@
...
-  BaseElements (const unsigned int n_compositional_fields)
-    :
-    compositional_fields (n_compositional_fields > 0 ? 3 : numbers::invalid_unsigned_int)
-  {}
-
+  void Introspection<dim>::free_finite_element_data ()
+  {
+    for (unsigned int i=0; i<fes.size(); ++i)
+      delete fes[i];
+    fes.clear();
```
Comment (bangerth) [inline source/simulator/introspection.cc]: Please also reset the multiplicities array.

**Comment by bangerth** [comment]
I'm not madly in love with the `free_*` function, but don't have a much better idea right now either. Let's just go with it.

Please address the two little things, rebase and merge yourself. Thanks for all the work, and my apologies for not getting to review it earlier :-(

**Pull request merged**